packages/devtools_flutter/lib/src/scaffold.dart

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/src/foundation/diagnostics.dart';
+
+import 'config.dart';
+import 'screen.dart';
+
+/// Scaffolding for a screen and navigation in the DevTools App
+///
+/// This widget will host Screen widgets.
+///
+/// [Config] defines what screens to show in the app.
+/// For a sample implementation, see [ConnectDevToolsScaffold].
+class DevToolsScaffold extends StatefulWidget {
+  const DevToolsScaffold({
+    Key key,
+    @required this.tabs,
+  })  : assert(tabs != null),
+        super(key: key);
+
+  /// A [Key] that indicates the page is showing in narrow-width mode.
+  static const Key narrowWidthKey = Key('Narrow Scaffold');
+
+  /// A [Key] that indicates the page is showing in full-width mode.
+  static const Key fullWidthKey = Key('Full-width Scaffold');
+
+  /// The width at or below which we treat the screen as narrow-width.
+  static const double narrowWidthThreshold = 800.0;
+
+  /// All of the screens that it's possible to navigate to from this Scaffold.
+  final List<Screen> tabs;
+
+  @override
+  State<StatefulWidget> createState() => DevToolsScaffoldState();
+}
+
+class DevToolsScaffoldState extends State<DevToolsScaffold>
+    with TickerProviderStateMixin {
+  /// A tag used for [Hero] widgets to keep the app title in the same place across page transitions.
+  static const String _titleTag = 'App Title';
+  TabController _controller;
+
+  @override
+  void initState() {
+    super.initState();
+    _setupController();
+  }
+
+  @override
+  void didUpdateWidget(DevToolsScaffold oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.tabs.length != oldWidget.tabs.length) {
+      var newIndex = 0;
+      if (_controller != null &&
+          widget.tabs.contains(oldWidget.tabs[_controller.index])) {
+        newIndex = widget.tabs.indexOf(oldWidget.tabs[_controller.index]);
+      }
+      _setupController();
+      _controller.index = newIndex;
+    }
+  }
+
+  @override
+  void dispose() {
+    _controller?.dispose();
+    super.dispose();
+  }
+
+  // Sets up the tab controllers.
```

Comment:
nit: r/controllers/controller

Reply:
The comment is redundant given the method name, so I've pulled it out.